s3access/reader.py

```diff
@@ -0,0 +1,114 @@
+import csv
+from abc import abstractmethod, ABC
+from io import BytesIO, StringIO
+from json import JSONDecoder, JSONDecodeError
+from typing import Generic, TypeVar, Union, Dict, Type, Iterator, Sequence, List
+
+import pandas as pd
+
+R = TypeVar('R')
+
+
+class Reader(ABC, Generic[R]):
+    @abstractmethod
+    def read(self, bs: Union[bytes, bytearray], columns: Dict[str, Type]) -> R:
+        raise NotImplementedError
+
+    @abstractmethod
+    def combine(self, results: Sequence[R]) -> R:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def serialization(self):
+        raise NotImplementedError
+
+    @property
+    def supports_caching(self):
+        return False
+
+    def read_cache(self, cache_file: str) -> R:
+        return NotImplemented
+
+    def write_cache(self, cache_file: str, contents: R):
+        return NotImplemented
+
+
+class Pandas(Reader[pd.DataFrame]):
+    def read(self, bs: Union[bytes, bytearray], columns: Dict[str, Type]) -> pd.DataFrame:
+        return pd.read_csv(BytesIO(bs), header=None, names=columns.keys(), dtype=columns)
+
+    def combine(self, results: Sequence[pd.DataFrame]) -> pd.DataFrame:
+        if not results:
+            return pd.DataFrame([])
```

Comment:
can never decide for this case if its better to return an empty dataframe (no columns and types) or simply None.

Reply:
The eternal question. Then the result should be `Optional[T]`.

I think an empty data frame is alright here, just as other implementations give an empty iterator, etc.